README.rst
===================================================
Community Detection HiDeF
===================================================

This repository creates a CDAPS compatible community detection Docker image using HiDeF `(Preprint) <https://doi.org/10.1101/2020.06.16.151555>`_
packaged from https://github.com/fanzheng10/HiDeF

Dependencies
------------

* `Docker <https://www.docker.com/>`_
* `make <https://www.gnu.org/software/make/>`_ (to build)
* Python (to build)

Direct invocation
------------------

Version `0.2.1` can be directly pulled from `Dockerhub <https://hub.docker.com/>`_ with this command:

.. code-block::

   docker pull coleslawndex/cdhidef:0.2.1

Building
--------

.. code-block::

   git clone https://github.com/idekerlab/cdhidef
   cd cdhidef
   make dockerbuild

Run **make** command with no arguments to see other build/deploy options including creation of Docker image

.. code-block::

   make

Output:

.. code-block::

   clean                remove all build, test, coverage and Python artifacts
   clean-build          remove build artifacts
   clean-pyc            remove Python file artifacts
   clean-test           remove test and coverage artifacts
   lint                 check style with flake8
   test                 run tests quickly with the default Python
   test-all             run tests on every Python version with tox
   coverage             check code coverage quickly with the default Python
   docs                 generate Sphinx HTML documentation, including API docs
   servedocs            compile the docs watching for changes
   testrelease          package and upload a TEST release
   release              package and upload a release
   dist                 builds source and wheel package
   install              install the package to the active Python's site-packages
   dockerbuild          build docker image and store in local repository
   dockerpush           push image to dockerhub


Usage
-----

.. code-block::

   docker run -v coleslawndex/cdhidef:0.2.1 -h


Credits
---------

This package was created with Cookiecutter_ and the `audreyr/cookiecutter-pypackage`_ project template.

.. _Cookiecutter: https://github.com/audreyr/cookiecutter
.. _`audreyr/cookiecutter-pypackage`: https://github.com/audreyr/cookiecutter-pypackage


HISTORY.rst
=======
History
=======

1.2.0 (2022-09-01)
------------------------------

* Added ``--convertresult`` which assumes the main
  input parameter points to HiDEF x.nodes or x.edges results
  file and parses that without running the HiDEF

1.1.1b1 (2020-01-15)
------------------------------

* Updated to pull cdhidef.1.1.1b1 of https://github.com/coleslaw481/HiDeF

* Added ``--numthreads`` flag with default value set to ``4``. This is to support
  new ``--numthreads`` flag in HiDeF 1.1.1 beta

* Renamed ``--ct`` flag to ``--p`` to match HiDeF 1.1.1 beta

* Reduced ``--maxres`` default to ``25.0``

* No longer passing ``--skipclug`` flag to HiDeF since HiDeF 1.1.1 beta
  removed this option

* Switched default algorithm ``--alg`` to ``leiden``

0.2.2 (2020-08-12)
------------------------------

* Updated to pull cdhidef.0.2.2 of https://github.com/coleslaw481/HiDeF
  which fixed a bug where --alg leiden was failing

0.2.1 (2020-06-15)
------------------------------

* Updated to pull cdhidef.0.2.1 of https://github.com/coleslaw481/HiDeF

0.2.0 (2020-05-19)
------------------------------

* Updated to pull cdhidef.0.2.0 of https://github.com/coleslaw481/HiDeF

0.1.1 (2020-05-14)
------------------------------

* Updated to generate new output that includes
  custom new column

0.0.1 (2020-03-26)
------------------

* First release


tests/test_cdhidef.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
test_cdhidef
----------------------------------

Tests for `cdhidef` module.
"""

import os
import sys
import unittest
import tempfile
import shutil
import io
import stat
import json
from unittest.mock import MagicMock
from cdhidef import cdhidefcmd


def write_fake_hidef(fakecmd, stdout, stderr, srcdatadir,
                     exitcode):
    """Fake hidef command script
    """
    f = open(fakecmd, 'w')
    f.write('#!/usr/bin/env python\n\n')
    f.write('import sys\n')
    f.write('import os\n')
    f.write('import shutil\n')
    f.write('import argparse\n\n')
    f.write('parser = argparse.ArgumentParser()\n')
    f.write('parser.add_argument(\'--o\')\n')
    f.write('theargs = parser.parse_known_args(sys.argv[1:])\n')
    f.write('sys.stdout.write(str(theargs) + \':::' +
            stdout + '\')\n')
    f.write('sys.stderr.write(\'' + stderr + '\')\n')

    f.write('try:\n')
    nodes_filename = cdhidefcmd.X_PREFIX + '.nodes'
    src_nodes_file = os.path.join(srcdatadir, nodes_filename)

    f.write('    shutil.copy(\'' + src_nodes_file +
            '\', theargs[0].o + \'.nodes\')\n')
    f.write('except Exception as e:\n')
    f.write('    sys.stderr.write(str(e))\n')

    f.write('try:\n')
    edges_filename = cdhidefcmd.X_PREFIX + '.edges'
    src_edges_file = os.path.join(srcdatadir, edges_filename)

    f.write('    shutil.copy(\'' + src_edges_file +
            '\', theargs[0].o + \'.edges\')\n')
    f.write('except Exception as e:\n')
    f.write('    sys.stderr.write(str(e))\n')
    f.write('sys.exit(' + str(exitcode) + ')\n')
    f.flush()
    f.close()
    os.chmod(fakecmd, stat.S_IRWXU)


class TestCdhidef(unittest.TestCase):

    TEST_DIR = os.path.dirname(__file__)

    HUNDRED_NODE_DIR = os.path.join(TEST_DIR, 'data',
                                    '100node_example')

    def setUp(self):
        pass

    def tearDown(self):
        pass

    def test_parse_args_all_defaults(self):
        myargs = ['inputarg']
        res = cdhidefcmd._parse_arguments('desc', myargs)
        self.assertEqual('inputarg', res.input)
        self.assertEqual(None, res.n)
        self.assertEqual(5, res.k)
        self.assertEqual(25.0, res.maxres)
        self.assertEqual(75, res.p)
        self.assertEqual('leiden', res.alg)
        self.assertEqual('hidef_finder.py', res.hidefcmd)
        self.assertEqual('/tmp', res.tempdir)

    def test_parse_args_custom_params(self):
        myargs = ['i2',
                  '--n', '1',
                  '--k', '3',
                  '--maxres', '0.6',
                  '--p', '8',
                  '--alg', 'leiden',
                  '--hidefcmd', 'foo',
                  '--tempdir', 'yo']
        res = cdhidefcmd._parse_arguments('desc', myargs)
        self.assertEqual('i2', res.input)
        self.assertEqual(1, res.n)
        self.assertEqual(3, res.k)
        self.assertEqual(0.6, res.maxres)
        self.assertEqual(8, res.p)
        self.assertEqual('leiden', res.alg)
        self.assertEqual('foo', res.hidefcmd)
        self.assertEqual('yo', res.tempdir)

    def test_build_optional_arguments_with_n(self):
        myargs = ['i2',
                  '--n', '1',
                  '--k', '3',
                  '--maxres', '0.6',
                  '--p', '8',
                  '--alg', 'leiden',
                  '--hidefcmd', 'foo',
                  '--tempdir', 'yo']
        res = cdhidefcmd._parse_arguments('desc', myargs)
        optargs = cdhidefcmd.build_optional_arguments(res)

        self.assertEqual('--n', optargs[0])
        self.assertEqual('1', optargs[1])

        self.assertEqual('--k', optargs[2])
        self.assertEqual('3', optargs[3])

        self.assertEqual('--maxres', optargs[4])
        self.assertEqual('0.6', optargs[5])

        self.assertEqual('--p', optargs[6])
        self.assertEqual('8', optargs[7])

        self.assertEqual('--alg', optargs[8])
        self.assertEqual('leiden', optargs[9])

    def test_build_optional_arguments(self):
        myargs = ['i2']
        res = cdhidefcmd._parse_arguments('desc', myargs)
        optargs = cdhidefcmd.build_optional_arguments(res)

        self.assertEqual('--k', optargs[0])
        self.assertEqual('5', optargs[1])

        self.assertEqual('--maxres', optargs[2])
        self.assertEqual('25.0', optargs[3])

        self.assertEqual('--p', optargs[4])
        self.assertEqual('75', optargs[5])

        self.assertEqual('--alg', optargs[6])
        self.assertEqual('leiden', optargs[7])

    def test_run_hidef_no_file(self):
        temp_dir = tempfile.mkdtemp()
        try:
            tfile = os.path.join(temp_dir, 'foo')
            myargs = [tfile]
            f_out = io.StringIO()
            f_err = io.StringIO()
            theargs = cdhidefcmd._parse_arguments('desc', myargs)
            res = cdhidefcmd.run_hidef(theargs, out_stream=f_out,
                                       err_stream=f_err)
            self.assertEqual(3, res)
            self.assertEqual(tfile + ' is not a file',
                             f_err.getvalue())
            self.assertEqual('', f_out.getvalue())
        finally:
            shutil.rmtree(temp_dir)

    def test_run_hidef_empty_file(self):
        temp_dir = tempfile.mkdtemp()
        try:
            tfile = os.path.join(temp_dir, 'foo')
            open(tfile, 'a').close()
            myargs = [tfile]
            f_out = io.StringIO()
            f_err = io.StringIO()
            theargs = cdhidefcmd._parse_arguments('desc', myargs)
            res = cdhidefcmd.run_hidef(theargs, out_stream=f_out,
                                       err_stream=f_err)
            self.assertEqual(4, res)
            self.assertEqual(tfile + ' is an empty file',
                             f_err.getvalue())
            self.assertEqual('', f_out.getvalue())
        finally:
            shutil.rmtree(temp_dir)

    def test_run_hidef_success(self):
        temp_dir = tempfile.mkdtemp()
        try:
            input_file = os.path.join(TestCdhidef.HUNDRED_NODE_DIR,
                                      'input.txt')

            f_out = io.StringIO()
            f_err = io.StringIO()
            fakecmd = os.path.join(temp_dir, 'foo.py')
            write_fake_hidef(fakecmd, 'stdout',
                             'stderr',
                             TestCdhidef.HUNDRED_NODE_DIR, 0)
            myargs = [input_file, '--tempdir', temp_dir,
                      '--hidefcmd', fakecmd]
            theargs = cdhidefcmd._parse_arguments('desc', myargs)

            res = cdhidefcmd.run_hidef(theargs, out_stream=f_out,
                                       err_stream=f_err)

            err_data = f_err.getvalue()

            self.assertTrue("'--skipgml', "
                            "'--k', '5', "
                            "'--maxres', "
                            "'25.0', '--p', '75', "
                            "'--alg', 'leiden'" in err_data)

            out_data = f_out.getvalue()
            self.assertEqual(4087, len(out_data))
            self.assertEqual(0, res)
        finally:
            shutil.rmtree(temp_dir)

    def test_run_hidef_success_altparams(self):
        temp_dir = tempfile.mkdtemp()
        try:
            input_file = os.path.join(TestCdhidef.HUNDRED_NODE_DIR,
                                      'input.txt')

            f_out = io.StringIO()
            f_err = io.StringIO()
            fakecmd = os.path.join(temp_dir, 'foo.py')
            write_fake_hidef(fakecmd, 'stdout',
                             'stderr',
                             TestCdhidef.HUNDRED_NODE_DIR, 0)
            myargs = [input_file,
                      '--n', '1',
                      '--k', '3',
                      '--maxres', '0.6',
                      '--p', '8',
                      '--alg', 'leiden',
                      '--hidefcmd', fakecmd,
                      '--tempdir', temp_dir]
            theargs = cdhidefcmd._parse_arguments('desc', myargs)

            res = cdhidefcmd.run_hidef(theargs, out_stream=f_out,
                                       err_stream=f_err)

            err_data = f_err.getvalue()

            self.assertTrue("'--skipgml', '--n', '1', "
                            "'--k', '3', "
                            "'--maxres', "
                            "'0.6', '--p', '8', "
                            "'--alg', 'leiden'" in err_data)

            out_data = f_out.getvalue()
            self.assertEqual(4087, len(out_data))
            self.assertEqual(0, res)
        finally:
            shutil.rmtree(temp_dir)

    def test_run_hidef_error(self):
        temp_dir = tempfile.mkdtemp()
        try:
            input_file = os.path.join(TestCdhidef.HUNDRED_NODE_DIR,
                                      'input.txt')

            f_out = io.StringIO()
            f_err = io.StringIO()
            fakecmd = os.path.join(temp_dir, 'foo.py')
            write_fake_hidef(fakecmd, 'stdout',
                             'stderr',
                             TestCdhidef.HUNDRED_NODE_DIR, 1)
            myargs = [input_file, '--tempdir', temp_dir,
                      '--hidefcmd', fakecmd]
            theargs = cdhidefcmd._parse_arguments('desc', myargs)

            res = cdhidefcmd.run_hidef(theargs, out_stream=f_out,
                                       err_stream=f_err)

            err_data = f_err.getvalue()

            self.assertTrue('Command failed with '
                            'non-zero exit code: 1 : ' in err_data)
            out_data = f_out.getvalue()
            self.assertEqual(0, len(out_data))
            self.assertEqual(1, res)
        finally:
            shutil.rmtree(temp_dir)

    def test_run_no_files_from_hidef(self):
        temp_dir = tempfile.mkdtemp()
        try:
            input_file = os.path.join(TestCdhidef.HUNDRED_NODE_DIR,
                                      'input.txt')

            f_out = io.StringIO()
            f_err = io.StringIO()
            fakecmd = os.path.join(temp_dir, 'foo.py')
            write_fake_hidef(fakecmd, 'stdout',
                             'stderr',
                             temp_dir, 0)
            myargs = [input_file, '--tempdir', temp_dir,
                      '--hidefcmd', fakecmd]
            theargs = cdhidefcmd._parse_arguments('desc', myargs)

            res = cdhidefcmd.run_hidef(theargs, out_stream=f_out,
                                       err_stream=f_err)

            err_data = f_err.getvalue()

            out_data = f_out.getvalue()
            self.assertEqual(0, len(out_data))
            self.assertTrue('Output from cmd: ' in err_data)
            self.assertEqual(5, res)
        finally:
            shutil.rmtree(temp_dir)

    def test_main_invalid_file(self):
        temp_dir = tempfile.mkdtemp()
        try:
            tfile = os.path.join(temp_dir, 'foo')
            myargs = ['prog', tfile]
            res = cdhidefcmd.main(myargs)
            self.assertEqual(3, res)
        finally:
            shutil.rmtree(temp_dir)

    def test_main_empty_file(self):
        temp_dir = tempfile.mkdtemp()
        try:
            tfile = os.path.join(temp_dir, 'foo')
            open(tfile, 'a').close()
            myargs = ['prog', tfile]
            res = cdhidefcmd.main(myargs)
            self.assertEqual(4, res)
        finally:
            shutil.rmtree(temp_dir)

    def test_createtmpdir(self):
        temp_dir = tempfile.mkdtemp()
        try:
            params = MagicMock()
            params.tempdir = temp_dir
            foo = cdhidefcmd.create_tmpdir(params)
            self.assertTrue(os.path.isdir(foo))
            self.assertEqual(temp_dir, os.path.dirname(foo))
        finally:
            shutil.rmtree(temp_dir)

    def test_convert_hidef_output_to_cdaps_with_hundred(self):
        f_out = io.StringIO()
        res = cdhidefcmd.convert_hidef_output_to_cdaps(f_out,
                                                       TestCdhidef.HUNDRED_NODE_DIR)
        self.assertEqual(None, res)
        self.assertEqual(4087, len(f_out.getvalue()))

    def test_convertresult_flag_set_with_nodes_file(self):
        f_out = io.StringIO()
        f_err = io.StringIO()
        input_file = os.path.join(TestCdhidef.HUNDRED_NODE_DIR, 'x.nodes')
        myargs = [input_file, '--convertresult']
        theargs = cdhidefcmd._parse_arguments('desc', myargs)

        res = cdhidefcmd.run_hidef(theargs, out_stream=f_out,
                                   err_stream=f_err)

        err_data = f_err.getvalue()

        out_data = f_out.getvalue()
        self.assertEqual(4087, len(out_data))
        heir_res = json.loads(out_data)
        with open(os.path.join(TestCdhidef.HUNDRED_NODE_DIR, 'x.expectedresult'), 'r') as f:
            expected_heir = json.load(f)
        self.assertEqual(expected_heir, heir_res)
        self.assertEqual('', err_data)
        self.assertEqual(0, res)

    def test_convertresult_flag_set_with_edges_file(self):
        f_out = io.StringIO()
        f_err = io.StringIO()
        input_file = os.path.join(TestCdhidef.HUNDRED_NODE_DIR, 'x.edges')
        myargs = [input_file, '--convertresult']
        theargs = cdhidefcmd._parse_arguments('desc', myargs)

        res = cdhidefcmd.run_hidef(theargs, out_stream=f_out,
                                   err_stream=f_err)

        err_data = f_err.getvalue()

        out_data = f_out.getvalue()
        self.assertEqual(4087, len(out_data))
        heir_res = json.loads(out_data)
        with open(os.path.join(TestCdhidef.HUNDRED_NODE_DIR, 'x.expectedresult'), 'r') as f:
            expected_heir = json.load(f)
        self.assertEqual(expected_heir, heir_res)
        self.assertEqual('', err_data)
        self.assertEqual(0, res)

    def test_convertresult_flag_set_with_invalid_suffix_file(self):
        f_out = io.StringIO()
        f_err = io.StringIO()
        input_file = os.path.join(TestCdhidef.HUNDRED_NODE_DIR, 'x.expectedresult')
        myargs = [input_file, '--convertresult']
        theargs = cdhidefcmd._parse_arguments('desc', myargs)

        res = cdhidefcmd.run_hidef(theargs, out_stream=f_out,
                                   err_stream=f_err)

        err_data = f_err.getvalue()

        out_data = f_out.getvalue()
        self.assertEqual(0, len(out_data))

        self.assertTrue('--convertresult expects input to be ' in err_data)
        self.assertEqual(6, res)

    def test_convertresult_flag_set_with_edges_file_missing(self):
        temp_dir = tempfile.mkdtemp()
        try:
            f_out = io.StringIO()
            f_err = io.StringIO()
            shutil.copy(os.path.join(TestCdhidef.HUNDRED_NODE_DIR, 'x.nodes'),
                        os.path.join(temp_dir, 'x.nodes'))
            input_file = os.path.join(temp_dir, 'x.nodes')
            myargs = [input_file, '--convertresult']
            theargs = cdhidefcmd._parse_arguments('desc', myargs)

            res = cdhidefcmd.run_hidef(theargs, out_stream=f_out,
                                       err_stream=f_err)

            err_data = f_err.getvalue()

            out_data = f_out.getvalue()
            self.assertEqual(3540, len(out_data))

            self.assertTrue('No such file' in err_data)
            self.assertEqual(5, res)
        finally:
            shutil.rmtree(temp_dir)


if __name__ == '__main__':
    sys.exit(unittest.main())


cdhidef/cdhidefcmd.py
#!/usr/bin/env python

import os
import sys
import argparse
import traceback
import subprocess
import uuid
import csv
import shutil

DEFAULT_ERR_MSG = ('Did not get any clusters from HiDeF. Not sure'
                   ' what is going on\n')

X_PREFIX = 'x'

CDRES_KEY_NAME = 'communityDetectionResult'

NODE_CX_KEY_NAME = 'nodeAttributesAsCX2'

ATTR_DEC_NAME = 'attributeDeclarations'

PERSISTENCE_COL_NAME = 'HiDeF_persistence'


class Formatter(argparse.ArgumentDefaultsHelpFormatter,
                argparse.RawDescriptionHelpFormatter):
    pass


def _parse_arguments(desc, args):
    """
    Parses command line arguments
    :param desc:
    :param args:
    :return:
    """
    parser = argparse.ArgumentParser(description=desc,
                                     formatter_class=Formatter)
    parser.add_argument('input',
                        help='Can be Edge file in tab delimited format or '
                             'if --convertresult is also set '
                             'this can be full path to xxx.nodes or '
                             'xxx.edges file '
                             'from existing hidef run')
    parser.add_argument('--n', type=int,
                        help='Target community number. Explore the'
                             'maximum resolution parameter until the '
                             'number of generated communities at this '
                             'resolution is close enough to this value. '
                             'Increase to get more smaller communities')
    parser.add_argument('--k', type=int, default=5,
                        help='Persistence threshold. Increase to '
                             'delete unstable clusters, and get fewer '
                             'communities')
    parser.add_argument('--maxres', type=float, default=25.0,
                        help='Maximum resolution parameter. '
                             'Increase to get more smaller communities')
    parser.add_argument('--p', default=75, type=int,
                        help='Consensus threshold. Threshold of '
                             'collapsing community graph and '
                             'choose gene as for each community')
    parser.add_argument('--alg', default='leiden',
                        choices=['louvain', 'leiden'],
                        help='Chose to use Louvain or the newer Leiden '
                             'algorithm. Must be "louvain" or "leiden"')
    parser.add_argument('--hidefcmd',
                        default='hidef_finder.py',
                        help='Path to hidef_finder.py command')
    parser.add_argument('--tempdir', default='/tmp',
                        help='Directory needed to hold files temporarily for processing')
    parser.add_argument('--csvmaxfieldsize', type=int, default=100000000,
                        help='Sets maximum field size for csv parser')
    parser.add_argument('--numthreads', type=int, default=4,
                        help='Sets number of threads to use. '
                             'If 0 or negative, then '
                             'value of multiprocessing.cpu_count() is used')
    parser.add_argument('--convertresult', action='store_true',
                        help='If set, it is assumed the <input> passed in is the '
                             'path to the .nodes or .edges file generated by hidef '
                             'and this program will just parse those results instead '
                             'of running hidef')
    return parser.parse_args(args)


def create_tmpdir(theargs):
    """
    Creates temp directory for hidef output with
    a unique name of format cdhidef_<UUID>

    :param theargs: Holds attributes from argparse
    :type theargs: `:py:class:`argparse.Namespace`
    :return: Path to temp directory
    :rtype: str
    """
    tmpdir = os.path.join(theargs.tempdir, 'cdhidef_' + str(uuid.uuid4()))
    os.makedirs(tmpdir, mode=0o755)
    return tmpdir


def run_hidef_cmd(cmd):
    """
    Runs hidef command as a command line process
    :param cmd_to_run: command to run as list
    :type cmd_to_run: list
    :return: (return code, standard out, standard error)
    :rtype: tuple
    """
    p = subprocess.Popen(cmd,
                         stdout=subprocess.PIPE,
                         stderr=subprocess.PIPE)

    out, err = p.communicate()

    return p.returncode, out, err


def get_max_node_id(nodes_file):
    """
    Examines the 'nodes_file' passed in and finds the value of
    highest node id.

    It is assumed the 'nodes_file' a tab delimited
    file of format:

    <CLUSTER NAME> <# NODES> <SPACE DELIMITED NODE IDS> <SCORE>

    :param nodes_file:
    :type nodes_file: Path to to nodes file from hidef output
    :return: highest node id found
    :rtype: int
    """
    maxval = None
    with open(nodes_file, 'r') as csvfile:
        linereader = csv.reader(csvfile, delimiter='\t')
        for row in linereader:
            for node in row[2].split(' '):
                if maxval is None:
                    maxval = int(node)
                    continue
                curval = int(node)
                if curval > maxval:
                    maxval = curval
    return maxval


def write_members_for_row(out_stream, row, cur_node_id):
    """
    Given a row from nodes file from hidef output the members
    of the clusters by parsing the <SPACE DELIMITED NODE IDS>
    as mentioned in :py:func:`#get_max_node_id` description.

    The output is written to `out_stream` for each node id
    in format:

    <cur_node_id>,<node id>,c-m;

    :param out_stream:
    :type out_stream: file like object
    :param row: Should be a line from hidef nodes file parsed
                by :py:func:`csv.reader`
    :type row: iterator
    :param cur_node_id: id of cluster that contains the nodes
    :type cur_node_id: int
    :return: None
    """
    for node in row[2].split(' '):
        out_stream.write(str(cur_node_id) + ',' +
                         node + ',c-m;')


def update_cluster_node_map(cluster_node_map, cluster, max_node_id):
    """
    Updates 'cluster_node_map' which is in format of

    <cluster name> => <node id>

    by adding 'cluster' to 'cluster_node_map' if it does not
    exist

    :param cluster_node_map: map of cluster names to node ids
    :type cluster_node_map: dict
    :param cluster: name of cluster
    :type cluster: str
    :param max_node_id: current max node id
    :type max_node_id: int
    :return: (new 'max_node_id' if 'cluster' was added otherwise 'max_node_id',
              id corresponding to 'cluster' found in 'cluster_node_map')
    :rtype: tuple
    """
    if cluster not in cluster_node_map:
        max_node_id += 1
        cluster_node_map[cluster] = max_node_id
        cur_node_id = max_node_id
    else:
        cur_node_id = cluster_node_map[cluster]
    return max_node_id, cur_node_id


def update_persistence_map(persistence_node_map, node_id, persistence_val):
    """

    :param persistence_node_map:
    :param node_id:
    :param persistence_val:
    :return:
    """
    if node_id not in persistence_node_map:
        persistence_node_map[node_id] = persistence_val


def write_communities(out_stream, edge_file, cluster_node_map):
    """
    Writes out links between clusters in COMMUNITYDETECTRESULT format
    as noted in :py:func:`#convert_hidef_output_to_cdaps`

    using hidef edge file set in 'edge_file' that is expected to
    be in this tab delimited format:

    <SOURCE CLUSTER> <TARGET CLUSTER> <default>

    This function converts the <SOURCE CLUSTER> <TARGET CLUSTER>
    to new node ids (leveraging 'cluster_node_map')

    and writes the following output:

    <SOURCE CLUSTER NODE ID>,<TARGET CLUSTER NODE ID>,c-c;

    to the 'out_stream'

    :param out_stream: output stream
    :type out_stream: file like object
    :param edge_file: path to hidef edges file
    :type edge_file: str
    :return: None
    """
    with open(edge_file, 'r') as csvfile:
        linereader = csv.reader(csvfile, delimiter='\t')
        for row in linereader:
            out_stream.write(str(cluster_node_map[row[0]]) + ',' +
                             str(cluster_node_map[row[1]]) + ',c-c;')
    out_stream.write('",')


def write_persistence_node_attribute(out_stream, persistence_map):
    """

    :param out_stream:
    :param persistence_map:
    :return:
    """
    out_stream.write('"' + NODE_CX_KEY_NAME + '": {')
    out_stream.write('"' + ATTR_DEC_NAME + '": [{')
    out_stream.write('"nodes": { "' + PERSISTENCE_COL_NAME +
                     '": { "d": "integer", "a": "p1", "v": 0}}}],')
    out_stream.write('"nodes": [')
    is_first = True
    for key in persistence_map:
        if is_first is False:
            out_stream.write(',')
        else:
            is_first = False
        out_stream.write('{"id": ' + str(key) + ',')
        out_stream.write('"v": { "p1": ' + str(persistence_map[key]) + '}}')

    out_stream.write(']}}')


def convert_hidef_output_to_cdaps(out_stream, tempdir, prefix=X_PREFIX):
    """
    Looks for x.nodes and x.edges in `tempdir` directory
    to generate output in COMMUNITYDETECTRESULT format:
    https://github.com/idekerlab/communitydetection-rest-server/wiki/COMMUNITYDETECTRESULT-format

    This method leverages

    :py:func:`#write_members_for_row`

    and

    :py:func:`#write_communities`

    to write output

    :param out_stream: output stream to write results
    :type out_stream: file like object
    :param tempdir:
    :type tempdir: str
    :return: None
    """
    nodefile = os.path.join(tempdir, prefix + '.nodes')
    max_node_id = get_max_node_id(nodefile)
    cluster_node_map = {}
    persistence_map = {}
    out_stream.write('{"communityDetectionResult": "')
    with open(nodefile, 'r') as csvfile:
        linereader = csv.reader(csvfile, delimiter='\t')
        for row in linereader:
            max_node_id, cur_node_id = update_cluster_node_map(cluster_node_map,
                                                               row[0],
                                                               max_node_id)
            update_persistence_map(persistence_map, cur_node_id, row[-1])
            write_members_for_row(out_stream, row,
                                  cur_node_id)
    edge_file = os.path.join(tempdir, prefix + '.edges')
    write_communities(out_stream, edge_file, cluster_node_map)
    write_persistence_node_attribute(out_stream, persistence_map)
    out_stream.write('\n')
    return None


def build_optional_arguments(theargs):
    """
    Builds list of optional arguments and values

    :param theargs: Holds attributes from argparse
    :type theargs: `:py:class:`argparse.Namespace`
    :return: arguments for hidef in list of str objects
    :rtype: list
    """
    cmdargs = []
    if theargs.n is not None:
        cmdargs.extend(['--n', str(theargs.n)])

    cmdargs.extend(['--k', str(theargs.k)])
    cmdargs.extend(['--maxres', str(theargs.maxres)])
    cmdargs.extend(['--p', str(theargs.p)])
    cmdargs.extend(['--alg', theargs.alg])
    cmdargs.extend(['--numthreads', str(theargs.numthreads)])
    return cmdargs


def run_hidef(theargs, out_stream=sys.stdout,
              err_stream=sys.stderr):
    """
    Runs hidef command line program and then converts
    output to CDAPS compatible format that is output to
    standard out.

    :param theargs: Holds attributes from argparse
    :type theargs: `:py:class:`argparse.Namespace`
    :param out_stream: stream for standard output
    :type out_stream: file like object
    :param err_stream: stream for standard error output
    :type err_stream: file like object
    :return: 0 upon success otherwise error
    :rtype: int
    """
    if theargs.input is None or not os.path.isfile(theargs.input):
        err_stream.write(str(theargs.input) + ' is not a file')
        return 3

    if os.path.getsize(theargs.input) == 0:
        err_stream.write(str(theargs.input) + ' is an empty file')
        return 4

    if theargs.convertresult is True:
        if not theargs.input.endswith('.nodes') and not theargs.input.endswith('.edges'):
            err_stream.write('--convertresult expects input to be .nodes or .edges output '
                             'file from hidef that was run outside this program')
            return 6
        rundir = os.path.dirname(theargs.input)
        prefix = os.path.basename(theargs.input)[0:-6]
        try:
            convert_hidef_output_to_cdaps(out_stream, rundir, prefix=prefix)
        except FileNotFoundError as fe:
            err_stream.write('No output from hidef: ' + str(fe) + '\n')
            return 5
        err_stream.flush()
        out_stream.flush()
        return 0

    cmdargs = []
    cmdargs.extend([theargs.hidefcmd, '--g', theargs.input,
                    '--skipgml'])
    cmdargs.extend(build_optional_arguments(theargs))

    csv.field_size_limit(theargs.csvmaxfieldsize)
    tmpdir = create_tmpdir(theargs)
    try:
        outval = os.path.join(tmpdir, X_PREFIX)
        cmdargs.extend(['--o', outval])
        err_stream.write('Running ' + str(' '.join(cmdargs)) + '\n')
        err_stream.flush()
        cmdecode, cmdout, cmderr = run_hidef_cmd(cmdargs)

        if cmdecode != 0:
            err_stream.write('Command failed with non-zero exit code: ' +
                             str(cmdecode) + ' : ' + str(cmderr) + '\n')
            return 1

        if len(cmdout) > 0:
            err_stream.write('Output from cmd: ' + str(cmdout) + '\n')

        if len(cmderr) > 0:
            err_stream.write('Error output from cmd: ' + str(cmderr) + '\n')

        try:
            convert_hidef_output_to_cdaps(out_stream, tmpdir)
        except FileNotFoundError as fe:
            err_stream.write('No output from hidef: ' + str(fe) + '\n')
            return 5
        return 0
    finally:
        err_stream.flush()
        out_stream.flush()
        shutil.rmtree(tmpdir)


def main(args):
    """
    Main entry point for program
    :param args: command line arguments usually :py:const:`sys.argv`
    :return: 0 for success otherwise failure
    :rtype: int
    """
    desc = """
    Runs HiDeF on command line, sending output to standard
    out in new 
    """
    theargs = _parse_arguments(desc, args[1:])
    try:
        return run_hidef(theargs, sys.stdout, sys.stderr)
    except Exception as e:
        sys.stderr.write('\n\nCaught exception: ' + str(e))
        traceback.print_exc()
        return 2


if __name__ == '__main__':  # pragma: no cover
    sys.exit(main(sys.argv))
